getter/cache.py

```diff
@@ -61,22 +76,33 @@ def update_cache(json_file_name, file_hash_str, file_identifier, file_type):
     con = sqlite3.connect(DATABASE_NAME)
     cur = con.cursor()
 
-    cur.execute(
-        """
-    INSERT INTO cache (original_file_name, hash, json_file)
-    VALUES (?, ?, ?)
-    ON CONFLICT(original_file_name) DO UPDATE SET hash=?
-    WHERE original_file_name=?
-    """,
-        (
-            file_identifier + "." + file_type,
-            file_hash_str,
-            file_identifier + ".json",
-            file_hash_str,
-            file_identifier + "." + file_type,
-        ),
-    )
-
-    con.commit()
-
-    shutil.copy(json_file_name, CACHE_DIR)
+    try:
+        cur.execute(
+            """
+        INSERT INTO cache (original_file_name, hash, json_file)
+        VALUES (?, ?, ?)
+        ON CONFLICT(original_file_name) DO UPDATE SET hash=?
+        WHERE original_file_name=?
+        """,
+            (
+                file_identifier + "." + file_type,
+                file_hash_str,
+                file_identifier + ".json",
+                file_hash_str,
+                file_identifier + "." + file_type,
+            ),
+        )
+        con.commit()
+        shutil.copy(json_file_name, CACHE_DIR)
+    except sqlite3.IntegrityError as e:
+        print(e)
+        print(f"Cache update error, removing {file_identifier} from cache")
+        cur.execute(
+            """
+        DELETE FROM cache where json_file=?
```

Comment:
If we delete this entry from the cache db, do we need to delete the actual json file too?

Reply:
After reading a bit more, I see this is called after creating the json file: https://github.com/ThreeSixtyGiving/datagetter/blob/384d1cd618cb65c546597c798fc9e82e7b4337d6/getter/get.py#L241
so deleting it would be counter-productive.

But can we not "fix" the cache entry, perhaps delete and re-create it with the new original file name, or update the hash/original file name?

Reply:
Thanks @R2ZER0 yeah that's a good suggestion. Two things that I wasn't confident about was 

1)  "Only a single ON CONFLICT clause, specifically the first ON CONFLICT clause with a matching conflict target, may run for each row of the INSERT. When an ON CONFLICT clause fires, all subsequent ON CONFLICT clauses are bypassed for that one row." 

2) By catching it as an exception I thought I might also be able to catch any other uniquness integrity errors in case I hadn't thought of them all. However now that we've made a cache fail non-fatal the risk of having missed some error case is probably mitigated.

What do you think?